app/views.py

```diff
@@ -1,37 +1,57 @@
 from app.forms import ProfileUpdateForm
-from app.models import Challenge, Profile, Project, SocialLinkAttachement
+from app.models import Challenge, Profile, Project, SocialLinkAttachement, Tag
 
 from django.contrib.auth import login
 from django.contrib.auth.forms import UserCreationForm
 from django.contrib.auth.mixins import LoginRequiredMixin, UserPassesTestMixin
+from django.contrib.contenttypes.models import ContentType
 from django.db.models import Q
 from django.forms.widgets import CheckboxSelectMultiple
 from django.urls import reverse_lazy
 from django.views.generic.base import TemplateView, ContextMixin
 from django.views.generic.edit import FormView, UpdateView
-
-from django.contrib.contenttypes.models import ContentType
-from django.shortcuts import get_object_or_404
+from django.shortcuts import get_object_or_404, redirect
 
 
 class IndexView(TemplateView):
     template_name = "explore.html"
 
+    def dispatch(self, request, *args, **kwargs):
+        if request.GET.get("type") not in ["challenge", "project", None]:
+            return redirect("index")
+        return super(IndexView, self).dispatch(request, *args, **kwargs)
+
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
-        if "type" not in self.request.GET or "q" not in self.request.GET:
-            context["challenges"] = Challenge.objects.all()[:3]
-            context["projects"] = Project.objects.all()[:9]
+
+        context["tags"] = Tag.objects.all()
+        context["selected_tags"] = []
+
+        if "type" not in self.request.GET or (
+            "q" not in self.request.GET and "tag" not in self.request.GET
+        ):
```

Comment:
This doesn't look great but I can't think of anything better.

Reply:
```python
        if not self.request.GET.get("type") or (
            not self.request.GET.get("q") and not self.request.GET.get("tag")
        ):
```

Is this preferable by any means?